gradio/routes.py

```diff
@@ -592,26 +588,29 @@ def robots_txt():
 ########
 
 
-def safe_join(directory: str, path: str) -> str | None:
-    """Safely path to a base directory to avoid escaping the base directory.
-    Borrowed from: werkzeug.security.safe_join"""
-    _os_alt_seps: List[str] = list(
-        sep for sep in [os.path.sep, os.path.altsep] if sep is not None and sep != "/"
-    )
+def is_relative_to(self, other):
+    # method exist on PurePath class from 3.9
+    try:
+        self.relative_to(other)
+        return True
+    except ValueError:
+        return False
 
-    if path != "":
-        filename = posixpath.normpath(path)
-    else:
-        return directory
 
-    if (
-        any(sep in filename for sep in _os_alt_seps)
-        or os.path.isabs(filename)
-        or filename == ".."
-        or filename.startswith("../")
-    ):
-        return None
-    return posixpath.join(directory, filename)
+def safe_join(directory: str, path: str) -> str:
+    if path == "":
+        raise HTTPException(400)
+
+    filename = Path(path)
+    root = Path(directory)
+    full_path = root.joinpath(filename).resolve()
+
+    if filename.is_absolute() or not is_relative_to(full_path.parent, directory):
```

Comment:
This cover Lines 609 to 611 from previous code.
- check if `path` is an absolute path
- avoid `path` accessing files outside of `directory`

Basically forbidden access, hence the 403